decisions/0003-cf-api-controller-repo.md

```diff
@@ -0,0 +1,89 @@
+# 3. CF API Controller Source Code Organization
+
+Date: 2020-06-17
+
+## Status
+
+In Review
+
+## Context
+
+Currently we store `capi-kpack-watcher` directly in `capi-k8s-release`. This has
+caused some complexity when bumping image versions in our concourse pipeline.
+We're also planning on reworking the watcher to use [kubebuilder](https://github.com/kubernetes-sigs/kubebuilder), and so now is a
+good time to change any source code organization decisions.
+
+Proposed locations for `capi-kpack-watcher` source code:
+
+1. `cf-api-kpack-watcher` a repo just for this particular component
+
+   1. Pros:
+
+      1. Very clear what goes where
+
+      1. It'd be a standard `kubebuilder` go repo
+
+      1. Fits our current pattern of go service repositories (1 repo for each service)
+
+   1. Cons:
+
+      1. Many more repos to manage to keep this pattern consistent in the
+       near-future as we will also create controllers for routes, droplets, etc
+
+1. `cf-api-controllers`/`cf-controllers` which would contain other future controllers we create for CF API resources
+
+   1. Pros:
+
+      1. It'd be a standard `kubebuilder` go repo with multiple controllers
```

Comment:
> It'd be a standard `kubebuilder` go repo

I think this is pretty valuable. One of the complaints from Rubyists familiar with Rails / other frameworks about `cloud_controller_ng` is that it violates conventions and is surprising. I've talked with folks at CF Summits who have said things like "I've been working with Rails for 3 years and couldn't figure out how to PR Cloud Controller."

I think it would be a lot friendlier to external contributions -- or even just new CAPIans -- if our k8s codebases adopted the conventions of the Kubernetes community.

Whether or not the Kubebuilder project layout is a strong convention in the community is another question though... 🤔

Reply:
Curious to understand if this is saying there is a single Kubebuilder project that would have multiple controllers or if there will be multiple Kubebuilder projects (one for each controller). From what I understand a Kubebuilder project can have multiple controllers, but they would live within the same binary, where as if you wanted multiple binaries for whatever reason you would need multiple Kubebuilder projects.

I think it would be considered non-standard to have multiple Kubebuilder projects in one repo because they would create their own scaffolding with its own `go.mod` file. That being said I don't consider it too hard to understand that a repo may have multiple Kubebuilder projects; probably just strange to see it at first glance.

This question probably also moves outside the realm of code organization at this point, but I'm curious because I've kind of been thinking about the same thing on our team (as we start work on ASGs) if we are going to be making multiple separate controllers or a bunch of controllers within the same Kubebuilder project.

All that said, I don't know if I have a strong opinion yet on either of these options and both are probably easy to understand, but I do lean towards a single Kubebuilder project with multiple controllers/single binary.

Reply:
As far as I have understood kubebuilder/controller-runtime one of the benefits of a single project with multiple controllers is that they all would share the same cache.

If that is beneficial or not of course depends on the overlap of resources the different controllers would be watching.

Reply:
It's also possible to have 2 separate controller-manager binaries in 1 kubebuilder project, it's just non standard. The standard pattern seems to be 1 binary, 1 process, 1 cache, many controllers, but if we need to break it down or componentize it in the future I don't think we'd have trouble with multiple binaries sharing 1 codebase.